### PR TITLE
Migrator Backfill Override

### PR DESCRIPTION
**Pull request by DaedalusG**
This PR introduces the ability to override the stitched-migration graph and define a migration to be backfilled at the time of the runner store instantiation.

**Why?**
We've been plagued by an old migration (`1648115472`) that continues to disappear and come back in many versions. You can learn more about this issue in the following PRs: https://github.com/sourcegraph/sourcegraph/pull/46969 https://github.com/sourcegraph/sourcegraph/pull/55650 which were previous attempts to resolve this issue.

The TL:DR of the issue is that during a multiversion upgrade we must unsquash (stitch) a stitched-migration graph together into a path for applying schema migrations during an multiversion upgrade, with breakpoints on minor versions to allow the execution of OOB migrations at the right time. In order to stitch this graph we need to determine which defined migrations may have been run prior to a squash and "backfill" those migrations to prevent them from being applied at the wrong time.

Migration `1648115472` is an old migration which was merged from a long lived branch after a squash operation had been performed on the migration graph. Despite several PRs to correct this migration via fixes to the backfill selection logic we continue to see this migration come back _(and fail)_ after new minor version releases. 

**How does the fix work**:
We bypass the stitching process [directly injecting](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@b229de696ec44ea426b59409172f29d12f712718/-/blob/internal/database/migration/store/store.go?L174) the migration as to be backfilled in the [instantiation](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@b229de696ec44ea426b59409172f29d12f712718/-/blob/internal/database/connections/live/runner.go?L55) of migrators handle on the databases the `store` and `runner`. 

**Alternative considered**:
The migrator codebase has some handles for dealing with problem migrations from the past in [rewriter.go](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@b229de696ec44ea426b59409172f29d12f712718/-/blob/internal/database/migration/stitch/rewriter.go?L102-115). This code area contains various rewrite functions applied during the generation for the stitched migration graph.

We may have used a method like this --
```go
func rewriteBackfillMigrations(schemaName string, _ oobmigration.Version, migrationIDs []int, contents map[string]string) {
    if schemaName != "frontend" {
        return
    }

    mapContents(contents, migrationFilename(1648115472, "up.sql"), func(_ string) string {
        return ``
    })
}
```
This function would overwrite the sql applied by migration `1648115472` giving it an empty string body and effectively making it irrelevant wether the migration was marked as backfilled or not during the production of the migration graph.

Ultimately this fix wasn't selected because it can only take effect after the release of a new minor version v5.4 in this case, and because there is less confidence in a method that my influence the generation of the migration graph in some way.

**Risks**:
Versions prior to the last migration graph squash performed in v4.41 attempting a multiversion upgrade may actually need to run this migration. Although in testing no early versions failed with the application of this change to the migrator it remains possible that some database in the wild may need to run this migration. _Although this is fairly unlikely_. In such cases an older version of migrator may be used with a stable upgrade path up to a version prior to v5.3.0

## Test plan

This was tested by running the upgradetest bazel release as shown below:
`VERSION=5.3.1 bazel run //testing/tools/upgradetest:sh_upgradetest_run --config darwin-docker --stamp --workspace_status_command=./dev/bazel_stamp_vars.sh -- all`

_note that this was originally a branch from https://github.com/sourcegraph/sourcegraph/pull/60594 which contains the testing automation to attempt a confirmation of the fix_

<details>

```
➜  sourcegraph git:(wg/migrator-backfill-rewriter) ✗ VERSION=5.3.1 bazel run //testing/tools/upgradetest:sh_upgradetest_run --config darwin-docker --stamp --workspace_status_command=./dev/bazel_stamp_vars.sh -- all
INFO: Analyzed target //testing/tools/upgradetest:sh_upgradetest_run (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //testing/tools/upgradetest:sh_upgradetest_run up-to-date:
  bazel-bin/testing/tools/upgradetest/sh_upgradetest_run
Aspect @@rules_rust//rust/private:clippy.bzl%rust_clippy_aspect of //testing/tools/upgradetest:sh_upgradetest_run up-to-date (nothing to build)
INFO: Elapsed time: 0.468s, Critical Path: 0.02s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Running command line: bazel-bin/testing/tools/upgradetest/sh_upgradetest_run testing/tools/upgradetest/upgradetest-darwin-arm64 cmd/migrator/image_tarball.sh cmd/frontend/image_tarball.sh internal/database/_codeinsights_squashed.sql internal/database/_codeintel_squashed.sql internal/database/_frontend_squashed.sql internal/database/_schema.codeinsights.json internal/database/_schema.codeinsights.md internal/database/_schema.codeintel.json internal/database/_schema.codeintel.md internal/database/_schema.json internal/database/_schema.md all
Loaded image: migrator:candidate
Loaded image: frontend:candidate
Latest stable release version:  5.3.0
Latest minor version:  5.3.0
Target version:  5.3.1
Migrator image used to upgrade:  migrator:candidate stamped as 5.3.1
Standard Versions: [5.2.0 5.2.1 5.2.2 5.2.3 5.2.4 5.2.5 5.2.6 5.2.7 5.3.0]
Multiversion Versions: [3.39.0 3.39.1 3.40.0 3.40.1 3.40.2 3.41.0 3.41.1 3.42.0 3.42.1 3.42.2 3.43.0 3.43.1 3.43.2 4.0.0 4.0.1 4.1.0 4.1.1 4.1.2 4.1.3 4.2.0 4.2.1 4.3.0 4.3.1 4.4.0 4.4.1 4.4.2 4.5.0 4.5.1 5.0.0 5.0.1 5.0.2 5.0.3 5.0.4 5.0.5 5.0.6 5.1.0 5.1.1 5.1.2 5.1.3 5.1.4 5.1.5 5.1.6 5.1.7 5.1.8 5.1.9]
Autoupgrade Versions: [3.39.0 3.39.1 3.40.0 3.40.1 3.40.2 3.41.0 3.41.1 3.42.0 3.42.1 3.42.2 3.43.0 3.43.1 3.43.2 4.0.0 4.0.1 4.1.0 4.1.1 4.1.2 4.1.3 4.2.0 4.2.1 4.3.0 4.3.1 4.4.0 4.4.1 4.4.2 4.5.0 4.5.1 5.0.0 5.0.1 5.0.2 5.0.3 5.0.4 5.0.5 5.0.6 5.1.0 5.1.1 5.1.2 5.1.3 5.1.4 5.1.5 5.1.6 5.1.7 5.1.8 5.1.9 5.2.0 5.2.1 5.2.2 5.2.3 5.2.4 5.2.5 5.2.6 5.2.7 5.3.0]
mvu:  3.39.0
std:  5.2.2
std:  5.2.0
std:  5.2.1
std:  5.2.5
std:  5.2.3
std:  5.2.7
std:  5.2.4
std:  5.3.0
std:  5.2.6
mvu:  3.39.1
mvu:  3.40.0
mvu:  3.40.1
mvu:  3.40.2
mvu:  3.41.0
mvu:  3.41.1
mvu:  3.42.0
mvu:  3.42.1
mvu:  3.42.2
mvu:  3.43.0
mvu:  3.43.1
mvu:  3.43.2
mvu:  4.0.0
mvu:  4.0.1
mvu:  4.1.0
mvu:  4.1.1
mvu:  4.1.2
mvu:  4.1.3
mvu:  4.2.0
mvu:  4.2.1
mvu:  4.3.0
mvu:  4.3.1
mvu:  4.4.0
mvu:  4.4.1
mvu:  4.4.2
mvu:  4.5.0
mvu:  4.5.1
mvu:  5.0.0
mvu:  5.0.1
mvu:  5.0.2
mvu:  5.0.3
mvu:  5.0.4
mvu:  5.0.5
mvu:  5.0.6
mvu:  5.1.0
mvu:  5.1.1
mvu:  5.1.2
mvu:  5.1.3
mvu:  5.1.4
mvu:  5.1.5
mvu:  5.1.6
mvu:  5.1.7
mvu:  5.1.8
mvu:  5.1.9
auto:  3.39.0
auto:  3.39.1
auto:  3.40.0
auto:  3.40.1
auto:  3.40.2
auto:  3.41.0
auto:  3.41.1
auto:  3.42.0
auto:  3.42.1
auto:  3.42.2
auto:  3.43.0
auto:  3.43.1
auto:  3.43.2
auto:  4.0.0
auto:  4.0.1
auto:  4.1.0
auto:  4.1.1
auto:  4.1.2
auto:  4.1.3
auto:  4.2.0
auto:  4.2.1
auto:  4.3.0
auto:  4.3.1
auto:  4.4.0
auto:  4.4.1
auto:  4.4.2
auto:  4.5.0
auto:  4.5.1
auto:  5.0.0
auto:  5.0.1
auto:  5.0.2
auto:  5.0.3
auto:  5.0.4
auto:  5.0.5
auto:  5.0.6
auto:  5.1.0
auto:  5.1.1
auto:  5.1.2
auto:  5.1.3
auto:  5.1.4
auto:  5.1.5
auto:  5.1.6
auto:  5.1.7
auto:  5.1.8
auto:  5.1.9
auto:  5.2.0
auto:  5.2.1
auto:  5.2.2
auto:  5.2.3
auto:  5.2.4
auto:  5.2.5
auto:  5.2.6
auto:  5.2.7
auto:  5.3.0
--- 🕵️  Standard Upgrade Tests:
✅ 5.2.0 Passed -- 1m17.25698625s
✅ 5.2.1 Passed -- 1m19.827390125s
✅ 5.2.2 Passed -- 1m18.722931334s
✅ 5.2.3 Passed -- 1m18.513840084s
✅ 5.2.4 Passed -- 1m20.015227459s
✅ 5.2.5 Passed -- 1m18.312065584s
✅ 5.2.6 Passed -- 1m18.505294708s
✅ 5.2.7 Passed -- 1m16.21996725s
✅ 5.3.0 Passed -- 1m16.139833625s
--- 🕵️  Multiversion Upgrade Tests:
✅ 3.39.0 Passed -- 1m30.079454875s
✅ 3.39.1 Passed -- 1m29.527892709s
✅ 3.40.0 Passed -- 1m29.576872041s
✅ 3.40.1 Passed -- 1m29.588949041s
✅ 3.40.2 Passed -- 1m28.668168708s
✅ 3.41.0 Passed -- 1m31.626891959s
✅ 3.41.1 Passed -- 1m28.8544895s
✅ 3.42.0 Passed -- 1m30.038192958s
✅ 3.42.1 Passed -- 1m30.000996417s
✅ 3.42.2 Passed -- 1m28.599743833s
✅ 3.43.0 Passed -- 1m26.437386667s
✅ 3.43.1 Passed -- 1m28.870884417s
✅ 3.43.2 Passed -- 1m28.968288709s
✅ 4.0.0 Passed -- 1m33.319444625s
✅ 4.0.1 Passed -- 1m32.977546959s
✅ 4.1.0 Passed -- 1m30.421875542s
✅ 4.1.1 Passed -- 1m31.079191459s
✅ 4.1.2 Passed -- 1m29.315826458s
✅ 4.1.3 Passed -- 1m28.153632417s
✅ 4.2.0 Passed -- 1m27.378116417s
✅ 4.2.1 Passed -- 1m19.56413325s
✅ 4.3.0 Passed -- 1m25.713798667s
✅ 4.3.1 Passed -- 1m24.336728292s
✅ 4.4.0 Passed -- 1m26.221845208s
✅ 4.4.1 Passed -- 1m25.023135s
✅ 4.4.2 Passed -- 1m25.734292875s
✅ 4.5.0 Passed -- 1m24.874949792s
✅ 4.5.1 Passed -- 1m25.695273625s
✅ 5.0.0 Passed -- 1m24.070825292s
✅ 5.0.1 Passed -- 1m21.077355291s
✅ 5.0.2 Passed -- 1m22.275236375s
✅ 5.0.3 Passed -- 1m14.823674875s
✅ 5.0.4 Passed -- 1m17.298590667s
✅ 5.0.5 Passed -- 1m16.995253375s
✅ 5.0.6 Passed -- 1m20.32679025s
✅ 5.1.0 Passed -- 1m36.00752775s
✅ 5.1.1 Passed -- 1m36.998838833s
✅ 5.1.2 Passed -- 1m36.742287417s
✅ 5.1.3 Passed -- 1m35.08960925s
✅ 5.1.4 Passed -- 1m34.378563041s
✅ 5.1.5 Passed -- 1m17.3475645s
✅ 5.1.6 Passed -- 1m22.750011209s
✅ 5.1.7 Passed -- 1m27.931231084s
✅ 5.1.8 Passed -- 1m27.058951458s
✅ 5.1.9 Passed -- 1m27.928285875s
--- 🕵️  Auto Upgrade Tests:
✅ 3.39.0 Passed -- 1m26.641972916s
✅ 3.39.1 Passed -- 1m32.649834625s
✅ 3.40.0 Passed -- 1m34.18229425s
✅ 3.40.1 Passed -- 1m33.321158208s
✅ 3.40.2 Passed -- 1m33.459369333s
✅ 3.41.0 Passed -- 1m31.791390416s
✅ 3.41.1 Passed -- 1m37.023741875s
✅ 3.42.0 Passed -- 1m38.5774185s
✅ 3.42.1 Passed -- 1m38.083377416s
✅ 3.42.2 Passed -- 1m39.572001708s
✅ 3.43.0 Passed -- 1m42.251591958s
✅ 3.43.1 Passed -- 1m39.155866959s
✅ 3.43.2 Passed -- 1m42.441260083s
✅ 4.0.0 Passed -- 1m41.9961375s
✅ 4.0.1 Passed -- 1m41.36466875s
✅ 4.1.0 Passed -- 1m39.594416792s
✅ 4.1.1 Passed -- 1m37.9531145s
✅ 4.1.2 Passed -- 1m35.161944375s
✅ 4.1.3 Passed -- 1m34.876948667s
✅ 4.2.0 Passed -- 1m30.633996667s
✅ 4.2.1 Passed -- 1m38.458033333s
✅ 4.3.0 Passed -- 1m29.396298834s
✅ 4.3.1 Passed -- 1m35.842782916s
✅ 4.4.0 Passed -- 1m32.785435708s
✅ 4.4.1 Passed -- 1m31.750538083s
✅ 4.4.2 Passed -- 1m33.028386208s
✅ 4.5.0 Passed -- 1m31.150549375s
✅ 4.5.1 Passed -- 1m28.409162458s
✅ 5.0.0 Passed -- 1m27.822578125s
✅ 5.0.1 Passed -- 1m25.242125542s
✅ 5.0.2 Passed -- 1m22.690952667s
✅ 5.0.3 Passed -- 1m19.791605959s
✅ 5.0.4 Passed -- 1m22.849635125s
✅ 5.0.5 Passed -- 1m22.135011792s
✅ 5.0.6 Passed -- 1m21.599056542s
✅ 5.1.0 Passed -- 1m31.257151667s
✅ 5.1.1 Passed -- 1m31.008227292s
✅ 5.1.2 Passed -- 1m32.027771834s
✅ 5.1.3 Passed -- 1m35.841212792s
✅ 5.1.4 Passed -- 1m33.965816167s
✅ 5.1.5 Passed -- 1m14.763225s
✅ 5.1.6 Passed -- 1m10.321258875s
✅ 5.1.7 Passed -- 1m9.856418584s
✅ 5.1.8 Passed -- 1m9.061232083s
✅ 5.1.9 Passed -- 1m7.533899875s
✅ 5.2.0 Passed -- 1m11.344999833s
✅ 5.2.1 Passed -- 1m6.140145292s
✅ 5.2.2 Passed -- 1m5.231158042s
✅ 5.2.3 Passed -- 1m4.166442584s
✅ 5.2.4 Passed -- 1m4.774428791s
✅ 5.2.5 Passed -- 1m4.475649583s
✅ 5.2.6 Passed -- 1m2.38947925s
✅ 5.2.7 Passed -- 1m1.748151667s
✅ 5.3.0 Passed -- 54.684385333s
```

</details>

### CODE REVIEW

File: testing/tools/upgradetest/README.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+The upgradetest folder contains the code for release tooling intended used to asses the stability of a new migrator and frontend build before tagging a new release. 
+The tooling takes the form of a cli interface and is intended to orchestrate a psuedo-release and upgrade using bazel built images. The tests are  inocations of our services relevant to schema migrations and versioning. 
+This test does not test Sourcegraph features, only the basic operations of database and frontend versioning.
+
+Commands are intended to be invoked using `bazil targets`, with each command executing tests intrended to be run in CI. The CI tests treat the current repo branch of `sourcegraph/sourcegraph` as a prospective release. 
+
+This "release branch" may be stamped with a version, or will be versioned `0.0.0+dev`. A stamped version of the release branch must be invoked with a certain bazel flags `--stamp`, `--workspace_status_command=./dev/bazel_stamp_vars.sh`,
+and additionally requires a `VERSION` env var to be set with a semantic version string `X.X.X`. 
+
+The general idea of the tests is to verify that a given upgrade process works as expected in a containerized end to end test.
```
[the remark this answers is not in the document]
Comment (jdpleiness) [reply testing/tools/upgradetest/README.md:11]: I agree with @BolajiOlajide, we should break the upgrade tests into their own PR with this just being the migrator fix.

File: internal/database/migration/store/store.go [hunk 2]
```diff
@@ -188,6 +193,12 @@ func (s *Store) BackfillSchemaVersions(ctx context.Context) error {
 	))
 }
 
+// backfillOverrides contains a list of migration IDs that are to be backfilled
+// after the stitched-migration graph has already been constructed.
+var backfillOverrides = []int64{
```
Comment (jdpleiness) [inline internal/database/migration/store/store.go:198]: I'd probably move this up in the file and declare it before we use it as well.

**Review by jdpleiness — COMMENTED** [review]
Huge thanks @DaedalusG for this ❤️

File: internal/database/migration/store/store.go [hunk 3]
```diff
@@ -171,7 +171,12 @@ func (s *Store) BackfillSchemaVersions(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
+
 	idsToBackfill := []int64{}
+	for _, id := range backfillOverrides {
+		idsToBackfill = append(idsToBackfill, int64(id))
+	}
```
[the remark this answers is not in the document]
Comment (DaedalusG) [reply internal/database/migration/store/store.go:178]: I think I'm going to leave this as is just for readability -- its more consistent with the other loop that is adding `idsToBackfill`

File: testing/tools/upgradetest/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (DaedalusG) [reply testing/tools/upgradetest/README.md:11]: Okayyyyyyyyy

**Pull request merged**